text/0000-extended-standard-library.md

```diff
@@ -0,0 +1,370 @@
+- Feature Name: Extended Standard Library
+- Start Date: 2025-05-08
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC describes an Extended Standard Library (ESL) for Rust. The ESL will consist of a set of crates providing commonly used functionality, subject to policies which promote excellent security, quality, and ease of use.
+
+# Motivation
+[motivation]: #motivation
+
+Rust programs, on average, have many dependencies. This is largely because the Rust standard library is intentionally limited in scope. As a consequence, Rust programs are currently required to pull in many third-party dependencies to get functionality that is considered table-stakes by 2025 standards.
+
+There are three issues this proposal aims to help address, all of which are related to the large number of independent entities providing crates that are widely depended on in Rust software:
+
+### 1. Impractically-sized web of trust
+
+Reliance on a high number of third-party dependencies creates a large web of trust for any non-trivial program. Given the size of this web of trust, it is practically impossible to be confident about the trustworthiness of the full dependency tree.
+
+Being a _trustworthy_ dependency means it:
+
+* Is developed and maintained by qualified, non-malicious authors
+* Has a responsible vulnerability disclosure policy
+* Has sufficient maintainer bandwidth to publish security updates in a reasonable timeframe
+* Has reasonable code review and testing standards
+* Uses secure authentication and release infrastructure
+* Depends only on libraries that are themselves trustworthy
+* Can be expected to maintain these properties well into the future
+
+Establishing confidence in the above characteristics for any individual dependency requires significant effort; establishing confidence in these for all of a project’s dependencies is impractical. As a result, many projects currently rely on fuzzy heuristics to determine the trustworthiness of a dependency, or even choose to ignore the problem altogether and "hope for the best."
+
+The status quo goes against Rust's goal of empowering everyone to build reliable software. Though it was never intended to be that way, as of today the Rust ecosystem nudges developers towards pulling in more dependencies than they can reasonably vet, which in the end hurts the reliability of the software.
```

Comment:
Do you have any numbers or other indications to demonstrate that the reliability of software has been hurt by the status quo? I simply don't accept the basic claim that decentralized code is bad and that we can solve things via centralizing. If anything, the rust project is one giant indication that centralization is a quick way to bury a small amount of people under a massive amount of work, until it all slows to a crawl because too many things are looked at by too few eyes.

Reply:
But is this RFC about centralisation of development, or is it about

1. ) giving users a seal of quality for some third party crates
2. ) providing funding to the maintainers of such approved crates

To me it reads like it's more of the latter.

Reply:
The entire thing is about moving approved crates into an approved github org so that only approved people can access any of it. The goal is a stamp of approval, the mechanism is centralization.

Reply:
The maintainers of the crates that move into the ESL would continue to be the maintainers. Anyone could contribute to the crates with PRs, same as before.

The GitHub org would enforce some best practices, particularly with regards to security, and additional resources would be available to help with maintenance. I think this is important for the ecosystem's most commonly used crates. Having them differ in terms of compliance with best practices, security policy and resources, and implementation of infrastructure and controls across many different GitHub orgs is not a good thing for Rust security, and it's probably more work in aggregate.

Reply:
So everything is the same, except when it's not the same? That seems incoherent.

I don't ask all this totally out of nothing. `tinyvec` is a crate that's available in the rust playground. I take this fact as a burden to be extremely careful of. How, concretely, does me moving my repo from my GitHub acocunt to your theoretical org meaningfully change anything for the better? Can you point to actual security situations that have occurred with rust that would have been solved if this RFC had been implemented before the event?

Reply:
To my knowledge there are ongoing initiatives to improve how surface security-information on crates.io. I don't know if it's a concrete goal but things like checking if a git commit matches published crate contents are things being discussed. This would benefit all crates, everyone, rather than just a single org.

> Having them differ in terms of compliance with best practices, security policy and resources, and implementation of infrastructure and controls across many different GitHub orgs is not a good thing for Rust security,

Our best practices vary a lot across crates. Popular ones get eyeballs, some maintenance-mode or "someone looks at it twice a year" dependencies can end up running on maintainer-doing-self-approvals.

Reply:
As far as I'm aware the actual publish step is basically always one person doing a publish, right? That seems like the point that we would want some RFC to put tight controls over.

It literally doesn't matter at all what code is in any git repository compared to how much it matters that the crates.io publication system is secured.

Reply:
I was also curious about the publish restrictions. Let's say a crate currently has two people with publish permissions. Then, that crate is promoted into the ESL. Do three people now have publish permissions?

Reply:
I think there's some work to make crates publishable from CI via ephemeral credentials?

But yeah, I'm just some pseudonymous rando on the internet who can and has published the libc crate with an unsecured token I had lying around on disk (an old artifact I promptly destroyed when I realized that it had become too powerful).

Reply:
Well currently in rust-lang/rust there's votes to merge a PR, but there's no way in crates.io to vote on a publication candidate before it's live in the system.

Some sort of team system that can review a proposed crates.io crate version before it becomes live to the public seems like an achievable goal for greatly increasing security on high use crates.

Reply:
I meant this one: https://github.com/rust-lang/crates.io/pull/11131
It's basically outsourcing the process management of creating a trusted release to github (and other trusted publishing platforms). Then the org can limit publishing to bots instead of humans and the bots will only publish after the appropriate reviews.

What imo would still be missing then is to compare what's on git and in the tarball so that nothing surprising snuck in due to a subverted CI process.

Reply:
Slippery slopes...

"Trusted Publishing"s name isn't meant to say other forms of publishing aren't trusted but to secure specifically focusing on the CI publishing process. The misleading name already caught on though.

There are trade offs with the current state of bot-based publishing such that we should not require only that. Without requiring bot-based publishing, you then need to allow direct pushes unless you make the release process onerous.

Reply:
> What imo would still be missing then is to compare what's on git and in the tarball so that nothing surprising snuck in due to a subverted CI process.

I personally would wish for quorums, here.

Automate the publication to crates.io, certainly, but then leave the crate in quarantine mode until other maintainers or dedicated auditors have vetted the published version.

Even small quorums (1 single independent auditor) would massively raise the bar compared to the statu quo.

(On projects where a single maintainer could also trigger a release by a bot, which is likely, you'd want a minimum quorum of 2, obviously)

Reply:
A) a trusted publishing process may already have a quorum on the CI side (github deployment environments can require multiple approvers), so reimplementing the same functionality on cratesio seems redundant.

B) If we're involving independent authors this seems like out of the hands of the maintainers, in that case "quarantine" would be the wrong framing imo. I think users being able to filter crates on `cargo add` and `cargo update` by whatever security policies they want to apply would be more appropriate than making it a global decision.
AIUI big orgs already run their own crate mirrors where they filter which crates can be pulled.

Reply:
Depending on how its done, quarantining packages could cause major problems for workspace maintainers.  It would help if we had a whole new publish API but that has been talked about for years.  Even once you have that, there can be reason to publish a workspace in fragments.